itests/org.openhab.core.io.net.tests/src/main/java/org/openhab/core/io/net/tests/ClientFactoryTest.java

```diff
@@ -0,0 +1,224 @@
+        }
+    }
+
+    private void beforeAll() throws Exception {
+        port = TestPortUtil.findFreePort();
+        wsUrl = "ws://" + HOST + ":" + port + "/ws";
+        httpUrl = "http://" + HOST + ":" + port + "/http1";
+        http2Url = "http://" + HOST + ":" + port + "/http2";
+        TestServer theServer = new TestServer(HOST, port);
+        theServer.startServer();
+        server = theServer;
+    }
+
+    private void testHttp1Client() throws Exception {
+        HttpClientFactory httpClientFactory = getService(HttpClientFactory.class);
+        assertNotNull(httpClientFactory);
+
+        HttpClient client = httpClientFactory.createHttpClient(CONSUMER);
+        // HttpClient client = new HttpClient();
+        assertNotNull(client);
+
+        try {
+            // start client
+            client.setConnectTimeout(WAIT_SECONDS * 1000);
+            client.start();
+
+            // send request
+            ContentResponse response = client.GET(httpUrl);
+
+            // check response
+            assertEquals(HttpStatus.OK_200, response.getStatus());
+            assertEquals(RESPONSE, response.getContentAsString());
+        } finally {
+            try {
+                // stop client
+                client.stop();
+            } catch (Exception e) {
+            }
+        }
+    }
+
+    private void testHttp2Client() throws Exception {
+        HttpClientFactory httpClientFactory = getService(HttpClientFactory.class);
+        assertNotNull(httpClientFactory);
+
+        HTTP2Client client = httpClientFactory.createHttp2Client(CONSUMER);
+        // HTTP2Client client = new HTTP2Client();
+        assertNotNull(client);
+
+        // initialize address
+        InetSocketAddress address = new InetSocketAddress(HOST, port);
+
+        Completable<@Nullable Session> sessionCompletable = new Completable<>();
+        Completable<@Nullable Stream> streamCompletable = new Completable<>();
+
+        TestStreamAdapter streamAdapter = new TestStreamAdapter();
+
+        MetaData.Request request = new MetaData.Request(HttpMethod.GET.toString(), new HttpURI(http2Url),
+                HttpVersion.HTTP_2, new HttpFields());
+
+        HeadersFrame headers = new HeadersFrame(request, null, true);
+
+        try {
+            // start client
+            client.setConnectTimeout(WAIT_SECONDS * 1000);
+            client.start();
+
+            // establish session
+            client.connect(address, new Session.Listener.Adapter(), sessionCompletable);
+            Session session = sessionCompletable.get(WAIT_SECONDS, TimeUnit.SECONDS);
+            assertNotNull(session);
+            assertFalse(session.isClosed());
+
+            // open stream
+            session.newStream(headers, streamCompletable, streamAdapter);
+            Stream stream = streamCompletable.get(WAIT_SECONDS, TimeUnit.SECONDS);
+            assertNotNull(stream);
+
+            // confirm stream is open
+            assertFalse(stream.isClosed());
```

Comment:
Why would the stream always be open here?

The `HeadersFrame` is created such that it should end the stream:

`HeadersFrame headers = new HeadersFrame(request, null, true);`

```java
    /**
     * <p>Creates a new {@code HEADERS} frame with an unspecified stream {@code id}.</p>
     * <p>The stream {@code id} will be generated by the implementation while sending
     * this frame to the other peer.</p>
     *
     * @param metaData the metadata containing HTTP request information
     * @param priority the PRIORITY frame associated with this HEADERS frame
     * @param endStream whether this frame ends the stream
     */
    public HeadersFrame(MetaData metaData, PriorityFrame priority, boolean endStream)
```

Now it isn't closed immediately either. But that may be an implementation detail. :wink: 

If I wait for it to be closed using:

`waitForAssert(() -> assertTrue(stream.isClosed()));`

it does get closed eventually.